[jit] Make CompilationUnit own Functions

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22207 [jit] refactor self to be a class again
* #22206 [jit] Give functions qualified names
* #22205 [jit] remove unused argument in import.cpp
* #22204 [jit] improvements to QualifiedName
* #22203 [jit] refactoring of module/object
* **#22202 [jit] Make CompilationUnit own Functions**

This PR begins the process of overhauling the ownership model for compilation
units. The first step is removing shared references to Functions. Now
CompilationUnit owns them all.

This also introduces StrongFunctionPtr, which owns the CU and a raw function
ptr ptr. This is because pybind requires a ref-counted wrapper for any
objects that go into Python.

There are a couple of hacks that are required because not all the pieces are in place.
1. Recursive scripting uses Python to own the functions. We don't qualify
function names so we can't just put them all in the global Python CU.
2. The FunctionValue sugared value is holding onto the CU to keep things alive.

Differential Revision: [D15998761](https://our.internmc.facebook.com/intern/diff/D15998761)

suo committed:
[jit] Make CompilationUnit own Functions

This PR begins the process of overhauling the ownership model for compilation
units. The first step is removing shared references to Functions. Now
CompilationUnit owns them all.

This also introduces StrongFunctionPtr, which owns the CU and a raw function
ptr ptr. This is because pybind requires a ref-counted wrapper for any
objects that go into Python.

There are a couple of hacks that are required because not all the pieces are in place.
1. Recursive scripting uses Python to own the functions. We don't qualify
function names so we can't just put them all in the global Python CU.

suo committed:
Update on "[jit] Make CompilationUnit own Functions"

[jit] Make CompilationUnit own Functions

This PR begins the process of overhauling the ownership model for compilation
units. The first step is removing shared references to Functions. Now
CompilationUnit owns them all.

This also introduces StrongFunctionPtr, which owns the CU and a raw function
ptr ptr. This is because pybind requires a ref-counted wrapper for any
objects that go into Python.

There are a couple of hacks that are required because not all the pieces are in place.
1. Recursive scripting uses Python to own the functions. We don't qualify
function names so we can't just put them all in the global Python CU.

gh-metadata: pytorch pytorch 22202 gh/suo/66/head

suo committed:
Update on "[jit] Make CompilationUnit own Functions"

[jit] Make CompilationUnit own Functions

This PR begins the process of overhauling the ownership model for compilation
units. The first step is removing shared references to Functions. Now
CompilationUnit owns them all.

This also introduces StrongFunctionPtr, which owns the CU and a raw function
ptr ptr. This is because pybind requires a ref-counted wrapper for any
objects that go into Python.

There are a couple of hacks that are required because not all the pieces are in place.
1. Recursive scripting uses Python to own the functions. We don't qualify
function names so we can't just put them all in the global Python CU.

gh-metadata: pytorch pytorch 22202 gh/suo/66/head

suo committed:
Update on "[jit] Make CompilationUnit own Functions"

[jit] Make CompilationUnit own Functions

This PR begins the process of overhauling the ownership model for compilation
units. The first step is removing shared references to Functions. Now
CompilationUnit owns them all.

This also introduces StrongFunctionPtr, which owns the CU and a raw function
ptr ptr. This is because pybind requires a ref-counted wrapper for any
objects that go into Python.

There are a couple of hacks that are required because not all the pieces are in place.
1. Recursive scripting uses Python to own the functions. We don't qualify
function names so we can't just put them all in the global Python CU.

gh-metadata: pytorch pytorch 22202 gh/suo/66/head

zdevito commented on torch/csrc/jit/passes/bailout_graph.cpp:
```diff
@@ -154,11 +154,12 @@ struct BailOutInserter {
-    auto unopt_func = graph_->create(prim::Constant);
-    unopt_func->output()->setType(FunctionType::create(std::move(func)));
-    unopt_func->insertBefore(*graph_->block()->nodes().begin());
+    auto unopt_func =
```
If this is planned to be in there for the long term (which seems reasonable because the original bailout graph is never executed directly so does not need to be a function), then we should give a name like prim::BailoutTemplate so it is more clear where it is being used.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -239,6 +239,15 @@ RegisterOperators reg(
+         "prim::GraphHolder() -> ()",
+         [](const Node* /* node */) {
+           return [](Stack& stack) {
+             // Returns an int so that we have an easy way to do graph traversal
```
This should throw.

suo replied:
Going to leave this as a follow up so that I don't muck around with the bottom of this stack too much. Added a TODO comment here in the code

zdevito commented on torch/csrc/jit/script/compilation_unit.h:
```diff
@@ -201,5 +207,20 @@ struct TORCH_API CompilationUnit {
+    TORCH_INTERNAL_ASSERT(function_);
+  }
+  std::shared_ptr<script::CompilationUnit> cu_;
+  Function* function_;
```
Should this class have operator* and operator->() implemented?

suo replied:
Since it's not supposed to be a widely used API I think we should avoid obfuscating what it really is behind operator overloads. If it turns out to be too clunky I can add.

zdevito commented on torch/csrc/jit/script/sugared_value.h:
```diff
@@ -268,7 +269,9 @@ struct FunctionValue : public SugaredValue {
 
  private:
-  std::shared_ptr<Function> callee_;
+  Function* callee_;
```
Can't these two things just be a StrongFunctionPtr?

suo replied:
We don't always need to construct an owning FunctionValue, and soon there will be owning FunctionValues. So I left them decomposed for now.

zdevito reviewed (approved):
Looks good as the first step.

WARNING: there are submodule diffs in here and foo.diff that should not be committed.

suo committed:
Update on "[jit] Make CompilationUnit own Functions"

[jit] Make CompilationUnit own Functions

This PR begins the process of overhauling the ownership model for compilation
units. The first step is removing shared references to Functions. Now
CompilationUnit owns them all.

This also introduces StrongFunctionPtr, which owns the CU and a raw function
ptr ptr. This is because pybind requires a ref-counted wrapper for any
objects that go into Python.

There are a couple of hacks that are required because not all the pieces are in place.
1. Recursive scripting uses Python to own the functions. We don't qualify
function names so we can't just put them all in the global Python CU.

gh-metadata: pytorch pytorch 22202 gh/suo/66/head

suo committed:
Update on "[jit] Make CompilationUnit own Functions"

[jit] Make CompilationUnit own Functions

This PR begins the process of overhauling the ownership model for compilation
units. The first step is removing shared references to Functions. Now
CompilationUnit owns them all.

This also introduces StrongFunctionPtr, which owns the CU and a raw function
ptr ptr. This is because pybind requires a ref-counted wrapper for any
objects that go into Python.

There are a couple of hacks that are required because not all the pieces are in place.
1. Recursive scripting uses Python to own the functions. We don't qualify
function names so we can't just put them all in the global Python CU.

gh-metadata: pytorch pytorch 22202 gh/suo/66/head
